blocks/soapy/test/qa_Soapy.cpp

```diff
@@ -0,0 +1,348 @@
+
+                device.setSampleRate(SOAPY_SDR_RX, 0, 1e6);
+                expect(approx(device.getSampleRate(SOAPY_SDR_RX, 0), 1e6, 1e2));
+            };
+
+            "time sources"_test = [&device] {
+                std::vector<std::string> timeSources = device.listAvailableTimeSources();
+                std::string              timeSource  = device.getTimeSource();
+                fmt::println("time sources: [{}] - active: '{}'", fmt::join(timeSources, ", "), timeSource);
+                if (!timeSources.empty()) {
+                    expect(std::ranges::find(timeSources, timeSource) != timeSources.cend());
+
+                    // device.getHardwareTime(SOAPY_SDR_RX, 0, activeAntenna);
+                    //  expect(eq(device.getAntenna(SOAPY_SDR_RX, 0), activeAntenna));
+                }
+                std::uint64_t hwTime = device.getHardwareTime();
+                fmt::println("time source hw time: {}", hwTime);
+            };
+
+            "simple acquisition"_test = [&device, &deviceDriver] {
+                using TValueType = std::complex<float>;
+                std::vector<TValueType> externalBuffer;
+                externalBuffer.resize(1UZ << 14UZ);
+
+                if (deviceDriver != "audio") {
+                    device.setSampleRate(SOAPY_SDR_RX, 0, 1'000'000.);
+                } else { // audio cannot go > 1 MS/s and other SDRs not below 1 MS/s
+                    device.setSampleRate(SOAPY_SDR_RX, 0, 44'100.);
+                }
+                device.setCenterFrequency(SOAPY_SDR_RX, 0, 107'000'000.);
+                Device::Stream rxStream = device.setupStream<TValueType, SOAPY_SDR_RX>();
+                rxStream.activate();
+
+                constexpr int maxSamples      = 100'000;
+                int           receivedSamples = 0UZ;
+                for (int i = 0; (i < 1000) && (receivedSamples <= maxSamples); ++i) {
+                    std::uint32_t pollingTimeOutUs = 10'000; // 10 ms
+                    int           flags            = 0;
+                    long long     time_ns          = 0;
+                    int           ret              = rxStream.readStream(flags, time_ns, pollingTimeOutUs, externalBuffer);
+                    if (ret > 0) {
+                        receivedSamples += ret;
+                    }
+                    if (i < 10 || ret > 0 || ret < -1) {
+                        fmt::print("{:3}: ", i);
+                        gr::blocks::soapy::detail::printSoapyReturnDebugInfo(ret, flags, time_ns);
+                    }
+                }
+                expect(ge(receivedSamples, 1000)) << fmt::format("did not received enough samples for deviceDriver = {}\n", deviceDriver);
+                rxStream.deactivate();
+            };
+
+            fmt::println("Basic API test - deviceDriver '{}' -- DONE", deviceDriver);
+        } //
+        | availableDeviceDriver;
+    //  |  std::array{"rtlsdr"s, "lime"s}; // alt: fixed tests
```

Comment:
It might be better to use predefined devices. In my tests, all available devices included some remote and audio devices that failed the tests.

Reply:
For me the test passed for 'rtlsdr', 'lime', and 'audio'. 

Some of the earlier failures I noticed where due to the drivers not implementing all the same capabilities (e.g. mutually exclusive sampling rates between audio and proper SDR devices, w/ vs. w/o gain setting capabilities). I'ts a pity that there is no default sw-only test-device that we could use in the CI ...

We should investigate more, if there are more/other devices that fail... thought that Soapy is a generic abstraction that would define either a lowest common denominator and/or provide dummy/empty implementations for those devices that cannot physically implement certain features.

If not, then this Soapy integration will be quite hard in terms of testing and guaranteeing that it works with any Soapy-wrapped SDR driver. 

Also the residual (static?) memory leak seems to be device/driver specific which seem to point to some leak in the soapy module implementation. For example, the same code shows more ASAN leaks for 'audio' than for 'rtlsdr' or 'lime'. :thinking: 

*N.B. need to find a solution, otherwise this will become a never-ending integration/testing issue....*